src/runtime/ndarray.cc

```diff
@@ -206,6 +211,46 @@ void NDArray::CopyFromTo(DGLArray* from, DGLArray* to) {
       from->dtype);
 }
 
+void NDArray::RecordedCopyFromTo(DGLArray* from, DGLArray* to, void* pyt_ctx) {
+  size_t from_size = GetDataSize(*from);
+  size_t to_size = GetDataSize(*to);
+  CHECK_EQ(from_size, to_size)
+      << "DGLArrayCopyFromTo: The size must exactly match";
+
+  CHECK(
+      from->ctx.device_type == to->ctx.device_type ||
+      from->ctx.device_type == kDGLCPU || to->ctx.device_type == kDGLCPU)
+      << "Can not copy across different ctx types directly";
+
+  // Use the context that is *not* a cpu context to get the correct device
+  // api manager.
+  DGLContext ctx = from->ctx.device_type != kDGLCPU ? from->ctx : to->ctx;
+  CHECK(ctx.device_type == kDGLCUDA)
+      << "Can not record event if not cuda device";
+  DeviceAPI::Get(kDGLCUDA)->RecordedCopyDataFromTo(
+      from->data, static_cast<size_t>(from->byte_offset), to->data,
+      static_cast<size_t>(to->byte_offset), from_size, from->ctx, to->ctx,
+      from->dtype, pyt_ctx);
+}
+
+NDArray NDArray::PinnedEmpty(
```

Comment:
Maybe better name than PinnedEmpty.
AllocEmptyPinnedMemory?

Reply:
There is a similar function `Empty` in the same file. `PinnedEmpty` is very similar to that function but the empty ndarray returned within pinned memory. The naming convention within `PinnedEmpty` function can also map to `Empty` function, i.e., `Empty` calls `AllocDataSpace`, while `PinnedEmpty` calls `AllocPinnedDataSpace`